### ISSUE TITLE
Test suite fails on big-endian machine

### ISSUE THREAD

**Issue by jasontibbitts** [cyrusimap/cyrus-imapd#1993]
I got to the point of doing a package build in Fedora's infrastructure, and in our development branch that builds for seven architectures.  Two of them (ppc64 and s390x/IBM System z) are big-endian.  The build succeeds, but make check doesn't.  Both fail with the same error:

```
Running unit tests
     CUnit - A unit testing framework for C - Version 2.1-3
     http://cunit.sourceforge.net/
Suite: annotate
  Test: getset_server_shared ...timeout: Unknown command '�'
```

That unprintable character is three bytes ("ef bf bd") but it could have been mangled in any number of ways by the time I gets to me.  You can see the complete build logs at https://koji.fedoraproject.org/koji/taskinfo?taskID=19726766 (which will be live for two weeks before the build is garbage collected).  Just click through the red links for the two failed builds and look for the "build.log" links.  "hw_info.log" gives useful information about the hardware.

I could just disable the test suite on those architectures, I guess, but while I suspect this is purely a test suite problem, it could also be a problem with the main code and so I'd like to make sure.

If you need access to either a ppc64 or s390x machine, please let me know.  Fedora does consider it important that everything supports LE and BE architectures and I can probably scrounge up someone to help if needed.  I've done some preliminary looking into this but haven't managed to figure it out.

**Comment by elliefm** [comment]
So it looks like it's barfing out on the very first test in the very first test suite :/

You can run the cunit test suites individually by making a target like check-_foo_ -- that is, to run just the annotate suite, run `make check-annotate`, or to run just the imapurl suite, `make check-imapurl` and so on.  Each of the files "cunit/_foo_.testc" constitutes a suite named _foo_

I don't have any particular leads at the moment, but I'm curious whether this issue presents for all, some, or just the annotate suite?

Also, the fact that you can get most of the Cassandane tests working is kind of implying to me at the moment that maybe the endianness issue is somewhere in our cunit infrastructure, rather than in Cyrus proper.  It'll be interesting to see if any of the cunit suites succeed, and which!

**Comment by elliefm** [comment]
For whatever it's worth:

```
cunit/timeout.c
144:            fprintf(stderr, "timeout: Unknown command '%c'\n", c);
```

**Comment by elliefm** [comment]
I don't suppose this teeny little patch fixes the issue?

```
diff --git a/cunit/timeout.c b/cunit/timeout.c
index f02c9ac..a444ac8 100644
--- a/cunit/timeout.c
+++ b/cunit/timeout.c
@@ -197,7 +197,7 @@ int timeout_init(void (*cb)(void))
 
 int timeout_begin(int millisec)
 {
-    int c;
+    char c;
     int r;
 
 // fprintf(stderr, "timeout_begin\n");
@@ -220,7 +220,7 @@ int timeout_begin(int millisec)
 
 int timeout_end(void)
 {
-    int c;
+    char c;
     int r;
 
 // fprintf(stderr, "timeout_end\n");
```

**Comment by jasontibbitts** [comment]
Well, it's better but it still doesn't pass.  Here's a build: https://koji.fedoraproject.org/koji/taskinfo?taskID=19734442

Just look at the ppc64 build and ignore the fact that some other architectures fail as well; they're failing in cassandane and not make check (ugh).  Maybe s390x will fail as well; we have fewer of those and the build hasn't even started yet.

To save you the clicking, here's the failed test output from ppc64. Everything else it runs appears to succeed, though I don't know if it runs all of the tests or stops at the first failure.  The segfault in there is kind of troubling, too.

```
Suite: conversations
  Test: open ...passed
  Test: getset ...FAILED
    1. ./cunit/conversations.testc:72  - CU_ASSERT_EQUAL(arrayu64_nth(&cids, 0)=0,C_CID=81985533797785328)
  Test: abort ...passed
  Test: prune ...FAILED
    1. ./cunit/conversations.testc:183  - CU_ASSERT_EQUAL(arrayu64_nth(&cids, 0)=0,C_CID1=73278643237678883)
    2. ./cunit/conversations.testc:189  - CU_ASSERT_EQUAL(arrayu64_nth(&cids, 0)=0,C_CID2=73579982724461108)
    3. ./cunit/conversations.testc:195  - CU_ASSERT_EQUAL(arrayu64_nth(&cids, 0)=0,C_CID3=73897814885606213)
    4. ./cunit/conversations.testc:203  - nseen >= 3
    5. ./cunit/conversations.testc:204  - ndeleted >= 2
    6. ./cunit/conversations.testc:205  - nseen - ndeleted >= 1
    7. ./cunit/conversations.testc:215  - CU_ASSERT_EQUAL(arrayu64_size(&cids)=1,0=0)
    8. ./cunit/conversations.testc:227  - CU_ASSERT_EQUAL(arrayu64_nth(&cids, 0)=0,C_CID3=73897814885606213)
  Test: two ...FAILED
    1. ./cunit/conversations.testc:312  - CU_ASSERT_EQUAL(arrayu64_nth(&cids, 0)=0,C_CID2=74779769797035367)
  Test: cid_encode ...passed
  Test: cid_decode ...passed
  Test: folder_rename ...FAILED
    1. ./cunit/conversations.testc:427  - CU_ASSERT_EQUAL(r=-1,0=0)
    2. ./cunit/conversations.testc:439  - CU_ASSERT_PTR_NOT_NULL_FATAL(conv)
  Test: folders .../bin/sh: line 10: 27303 Segmentation fault      (core dumped) $vg ./unit $f
make[3]: Leaving directory '/builddir/build/BUILD/cyrus-imapd-3.0.1'
make[3]: *** [Makefile:6559: check-local] Error 139
make[2]: Leaving directory '/builddir/build/BUILD/cyrus-imapd-3.0.1'
make[2]: *** [Makefile:6325: check-am] Error 2
make[1]: Leaving directory '/builddir/build/BUILD/cyrus-imapd-3.0.1'
make[1]: *** [Makefile:6033: check-recursive] Error 1
RPM build errors:
make: *** [Makefile:6327: check] Error 2
```

**Comment by elliefm** [comment]
That looks useful, cheers, will see if I can sort that one out too.  One step at a time!

**Comment by elliefm** [comment]
This doesn't bode well, the first failure (line 72) is from this:

```
    /* get should still succeed after the db is closed & reopened */
    memset(&cid, 0x45, sizeof(cid));
    arrayu64_truncate(&cids, 0);
    r = conversations_get_msgid(state, C_MSGID, &cids);
    CU_ASSERT_EQUAL(r, 0);
    CU_ASSERT_EQUAL(arrayu64_nth(&cids, 0), C_CID);
```

There doesn't seem to be anything endianness-dependent in the test itself, which suggests that the conversations module itself is busted -- doh!

**Comment by elliefm** [comment]
Okay, I don't expect this to _fix_ this current problem, but I'm hoping the cunit conversations tests will report more precise errors with the following patch applied, which in turn might help narrow down where the issue lies.  Can you give it a shot please?  Cheers

```
diff --git a/imap/conversations.c b/imap/conversations.c
index 504c5db..2b151dc 100644
--- a/imap/conversations.c
+++ b/imap/conversations.c
@@ -560,11 +560,14 @@ EXPORTED int conversations_get_msgid(struct conversations_state *state,
                       &data, &datalen,
                       &state->txn);
 
+    if (r == CYRUSDB_NOTFOUND)
+        return 0; /* not an error, but nothing more to do */
+
     if (!r) r = _conversations_parse(data, datalen, cids, NULL);
 
     if (r) arrayu64_truncate(cids, 0);
 
-    return 0;
+    return r;
 }
 
 /*
```

**Comment by jasontibbitts** [comment]
Sorry for the late response; Monday was a holiday in the US.  A fresh build is at https://koji.fedoraproject.org/koji/taskinfo?taskID=19785301

Both ppc64 and s390x fail pretty much identically; here's the relevant portion of the log.  I'm not sure there's actually any more useful information in there, though.

```
Suite: conversations
  Test: open ...passed
  Test: getset ...FAILED
    1. ./cunit/conversations.testc:71  - CU_ASSERT_EQUAL(r=-1,0=0)
    2. ./cunit/conversations.testc:72  - CU_ASSERT_EQUAL(arrayu64_nth(&cids, 0)=0,C_CID=81985533797785328)
  Test: abort ...passed
  Test: prune ...FAILED
    1. ./cunit/conversations.testc:182  - CU_ASSERT_EQUAL(r=-1,0=0)
    2. ./cunit/conversations.testc:183  - CU_ASSERT_EQUAL(arrayu64_nth(&cids, 0)=0,C_CID1=73278643237678883)
    3. ./cunit/conversations.testc:188  - CU_ASSERT_EQUAL(r=-1,0=0)
    4. ./cunit/conversations.testc:189  - CU_ASSERT_EQUAL(arrayu64_nth(&cids, 0)=0,C_CID2=73579982724461108)
    5. ./cunit/conversations.testc:194  - CU_ASSERT_EQUAL(r=-1,0=0)
    6. ./cunit/conversations.testc:195  - CU_ASSERT_EQUAL(arrayu64_nth(&cids, 0)=0,C_CID3=73897814885606213)
    7. ./cunit/conversations.testc:203  - nseen >= 3
    8. ./cunit/conversations.testc:204  - ndeleted >= 2
    9. ./cunit/conversations.testc:205  - nseen - ndeleted >= 1
    10. ./cunit/conversations.testc:215  - CU_ASSERT_EQUAL(arrayu64_size(&cids)=1,0=0)
    11. ./cunit/conversations.testc:220  - CU_ASSERT_EQUAL(r=-1,0=0)
    12. ./cunit/conversations.testc:226  - CU_ASSERT_EQUAL(r=-1,0=0)
    13. ./cunit/conversations.testc:227  - CU_ASSERT_EQUAL(arrayu64_nth(&cids, 0)=0,C_CID3=73897814885606213)
  Test: two ...FAILED
    1. ./cunit/conversations.testc:311  - CU_ASSERT_EQUAL(r=-1,0=0)
    2. ./cunit/conversations.testc:312  - CU_ASSERT_EQUAL(arrayu64_nth(&cids, 0)=0,C_CID2=74779769797035367)
  Test: cid_encode ...passed
  Test: cid_decode ...passed
  Test: folder_rename ...FAILED
    1. ./cunit/conversations.testc:427  - CU_ASSERT_EQUAL(r=-1,0=0)
    2. ./cunit/conversations.testc:439  - CU_ASSERT_PTR_NOT_NULL_FATAL(conv)
  Test: folders .../bin/sh: line 10: 23048 Segmentation fault      (core dumped) $vg ./unit $f
make[3]: *** [Makefile:6559: check-local] Error 139
make[3]: Leaving directory '/builddir/build/BUILD/cyrus-imapd-3.0.1'
make[2]: *** [Makefile:6325: check-am] Error 2
make[2]: Leaving directory '/builddir/build/BUILD/cyrus-imapd-3.0.1'
make[1]: *** [Makefile:6033: check-recursive] Error 1
make[1]: Leaving directory '/builddir/build/BUILD/cyrus-imapd-3.0.1'
make: *** [Makefile:6327: check] Error 2
```

s390x fails identically.

**Comment by elliefm** [comment]
Nah that's helpful!  sNotice this:

```
    1. ./cunit/conversations.testc:71  - CU_ASSERT_EQUAL(r=-1,0=0)
    2. ./cunit/conversations.testc:72  - CU_ASSERT_EQUAL(arrayu64_nth(&cids, 0)=0,C_CID=81985533797785328)
```

We can now see that the failure is actually occurring one line earlier, and the arrayu64_nth(...) failures are a natural consequence of some problem in the prior call :)

I think line 71, `r=-1` means it has thrown `CYRUSDB_IOERROR = -1`.

Will keep digging and see what next step presents itself.

**Comment by elliefm** [comment]
Okay I've dug around in the conversations API and it really does look like that error being handed down from cyrusdb.  But we don't get to see the results of the cyrusdb tests because they're alphabetically sorted such that they run _after_ the conversations tests, doh!

I've changed the order that the tests run so that the database tests are run first -- e3ee582fcf74e0754d2a850fbabe270125e334e8 is the relevant commit for the 3.0 branch.  Are you able to put through another test run with this commit applied?  With it, I expect something in the (newly renamed) aaa-db suite to fail -- which will hopefully point us towards the next step!

Thanks so much for working through this with me, it's really great to be able to exercise these code corners on platforms we don't usually have access to :)

**Comment by elliefm** [comment]
I've also added 443e546 to the 3.0 branch which corrects the conversations suite to treat failures to open the conversations database as fatal errors. I think this should prevent it segfaulting out later on, so hopefully we get to see some additional tests/suites being run!

**Comment by jasontibbitts** [comment]
OK, a build is at https://koji.fedoraproject.org/koji/taskinfo?taskID=19788126

ppc64 log is at https://kojipkgs.fedoraproject.org//work/tasks/8131/19788131/build.log

There is now way too much output from cunit for me to paste here; search for "Suite:" in the build log to go directly to the relevant bits.

**Comment by elliefm** [comment]
So there's thousands of errors from cyrusdb, specifically all from twoskip.  They seem to follow one of two patterns:

* fetch after commit
* fetch after close/reopen

I'm inclined to suspect something isn't being cleaned up properly and is then getting reused badly, but I can't see anything obviously like that.  And the correlation with endianness is weird.  @brong, do you have any thoughts here?

**Comment by jasontibbitts** [comment]
If it would help to have direct access to one of these machines for you to do some testing, I can try to arrange that.  Just let me know. Unfortunately I'm kind of on a deadline here if I want to make Fedora 26. While there s390x is not a primary arch, ppc64 is so I can't build for that release either.

**Comment by jasontibbitts** [comment]
I also wanted to add that if I just skip the 'make check' call and proceed to running Cassandane, the result isn't pretty.  Almost all of the Cassandane tests error out with things like:
```Perl exception: Cannot setacl for user.cassandane: IMAP Command : 'setacl' failed. Response was : no - System I/O error```
which looks pretty bad.  Others fail because they found stray processes.  I'm not sure what's a Cassandane problem and what's a problem with the core code at this point, but things in general just don't look good on big-endian machines right now.

**Comment by elliefm** [comment]
Yeah, whatever's broken in that twoskip cyrusdb handling is going to make pretty much everything in cassandane fail really hard, because twoskip is the default database format :(

**Comment by elliefm** [comment]
Access to a system might be useful.  At the moment I haven't even the beginnings of a hunch as to where things start going awry, maybe being able to step through in a debugger would help

**Comment by elliefm** [comment]
Have traced the twoskip problem down to a crc mismatch during `relocate()`...

Looking ahead the `crc32` suite itself has a bunch of failing tests, so maybe the problem is in our crc32 code, rather than in twoskip itself.

**Comment by elliefm** [comment]
For reference: 

```
Suite: crc32
  Test: rfc3720 ...FAILED
    1. ./cunit/crc32.testc:67  - CU_ASSERT_EQUAL(c=1728196545,INCRCRC=1188919630)
    2. ./cunit/crc32.testc:70  - CU_ASSERT_EQUAL(c=820022739,DECRCRC=289397596)
    3. ./cunit/crc32.testc:73  - CU_ASSERT_EQUAL(c=430929213,SCSIREADCRC=3650501206)
  Test: crc32c ...FAILED
    1. ./cunit/crc32.testc:92  - CU_ASSERT_EQUAL(c=1011723160,CRC32C=3753174729)
    2. ./cunit/crc32.testc:95  - CU_ASSERT_EQUAL(c=3337013816,CRC32C=3753174729)
  Test: map ...FAILED
    1. ./cunit/crc32.testc:105  - CU_ASSERT_EQUAL(c=1086647694,CRC32=1926722702)
  Test: iovec ...passed
  Test: zero ...passed
  Test: one ...passed
  Test: nine ...FAILED
    1. ./cunit/crc32.testc:153  - CU_ASSERT_EQUAL(c=1328763230,CRC32=3421780262)
  Test: alpha ...FAILED
    1. ./cunit/crc32.testc:163  - CU_ASSERT_EQUAL(c=1401146484,CRC32=1277644989)
  Test: null ...passed
  Test: twonull ...passed
  Test: bigbuf ...FAILED
    1. ./cunit/crc32.testc:198  - CU_ASSERT_EQUAL(c=3859746957,CRC32=2798229442)
  Test: iovec_empty ...passed
  Test: iovec_blank ...passed

Run Summary:    Type  Total    Ran Passed Failed Inactive
              suites     40      1    n/a      0        0
               tests    437     13      7      6        0
             asserts     18     18      9      9      n/a
```

**Comment by elliefm** [comment]
Looks like it's getting into the `crc32c_sw()` implementation, which is here: https://github.com/cyrusimap/cyrus-imapd/blob/39abcbb5e5a83fabcbc1604f0babf72646a08e57/lib/crc32c.c#L609-L641

This is pretty opaque to me :(

**Comment by jasontibbitts** [comment]
I would immediately wonder if LITTLE_ENDIAN is being defined properly (or not defined, as the case may be).  Just got up but I'll check some logs.

**Comment by jasontibbitts** [comment]
OK, I applied the following patch and did a fresh build.  I also installed a couple of extra cyrus-sasl packages to get rid of some test suite failures related to those.  The result?  make check passes!

The only weirdness is in the output is three instances of this one line:


```
Suite: libconfig
  Test: int ...passed
  Test: string ...passed
  Test: switch ...passed
  Test: stringlist_value ...passed
  Test: stringlist_default ...passed
  Test: stringlist_invalid ...
unit: fatal(invalid value 'junk' for annotation_db in line 2)
passed
  Test: enum_value ...passed
  Test: enum_default ...passed
  Test: enum_invalid ...
unit: fatal(invalid value 'junk' for delete_mode in line 2)
passed
  Test: bitfield_value ...passed
  Test: bitfield_default ...passed
  Test: bitfield_invalid ...
unit: fatal(invalid value 'junk' for httpmodules in line 2)
passed
```

```
diff --git a/lib/crc32.c b/lib/crc32.c
index c42e7d6..4193f0f 100644
--- a/lib/crc32.c
+++ b/lib/crc32.c
@@ -611,7 +611,7 @@ static const uint32_t crc32_lookup[16][256] =
     }
 };

-#ifndef LITTLE_ENDIAN
+#if BYTE_ORDER != LITTLE_ENDIAN
 /* swap endianness */
 static uint32_t swap(uint32_t x)
 {
@@ -634,7 +634,7 @@ static uint32_t crc32_slice8(uint32_t prev, const void *data, size_t length)

     /* process eight bytes at once (Slicing-by-8) */
     while (length >= 8) {
-#ifdef LITTLE_ENDIAN
+#if BYTE_ORDER == LITTLE_ENDIAN
         uint32_t one = *current++ ^ crc;
         uint32_t two = *current++;
         crc = crc32_lookup[0][(two>>24) & 0xFF] ^
@@ -682,7 +682,7 @@ static uint32_t crc32_slice16(uint32_t prev, const void *data, size_t length)
     while (length >= bytes_at_once) {
         size_t unrolling;
         for (unrolling = 0; unrolling < unroll; unrolling++) {
-#ifdef LITTLE_ENDIAN
+#if BYTE_ORDER == LITTLE_ENDIAN
         uint32_t one   = *current++ ^ crc;
         uint32_t two   = *current++;
         uint32_t three = *current++;
diff --git a/lib/crc32c.c b/lib/crc32c.c
index e33e793..66fc16e 100644
--- a/lib/crc32c.c
+++ b/lib/crc32c.c
@@ -591,7 +591,7 @@ static const uint32_t crc32c_lookup[4][256] = {
     },
 };

-#ifndef LITTLE_ENDIAN
+#if BYTE_ORDER != LITTLE_ENDIAN
 /* swap endianness */
 static uint32_t swap(uint32_t x)
 {
@@ -615,7 +615,7 @@ static uint32_t crc32c_sw(uint32_t crc, const void *buf, size_t len) {

     /* process four bytes at once (slicing-by-4) */
     while (len >= 4) {
-#ifdef LITTLE_ENDIAN
+#if BYTE_ORDER == LITTLE_ENDIAN
         crc1 = *cur++ ^ crc0;
         crc0 = crc32c_lookup[0][(crc1>>24) & 0xff] ^
                crc32c_lookup[1][(crc1>>16) & 0xff] ^
diff --git a/lib/xsha1.c b/lib/xsha1.c
index 1c4547e..6e41eee 100644
--- a/lib/xsha1.c
+++ b/lib/xsha1.c
@@ -89,7 +89,7 @@ struct _SHA_CTX {
 /* blk0() and blk() perform the initial expand. */
 /* I got the idea of expanding during the round function from SSLeay */

-#ifdef LITTLE_ENDIAN
+#if BYTE_ORDER == LITTLE_ENDIAN
 #define blk0(i) (block->l[i] = (rol(block->l[i],24)&(sha1_quadbyte)0xFF00FF00) \
     |(rol(block->l[i],8)&(sha1_quadbyte)0x00FF00FF))
 #else
```

**Comment by elliefm** [comment]
Oh awesome :D

**Comment by elliefm** [comment]
Patched in master and 3.0 :)

> The only weirdness is in the output is three instances of this one line:
> 
> `unit: fatal(invalid value 'junk' for annotation_db in line 2)`

That's specifically testing for that fatal error in that case.  Bit annoying that it whinges about it out loud, not sure if there's much we can do, but that `fatal()` is the test succeeding: https://github.com/cyrusimap/cyrus-imapd/blob/c2fdcba2f3a562261d3db09ae86a54e34d6aa96a/cunit/libconfig.testc#L177-L185

**Comment by jasontibbitts** [comment]
OK, glad to see the end of this one.
